File: src/arlo_controller_dmaking/src/SimulationController.cpp
/*
 * SimulationController.cpp
 *
 *  Created on: Apr 14, 2020
 *      Author: antonio
 */

#include "SimulationController.h"
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

SimulationController::SimulationController(double maxSTime, int tRate) : linear_{0},
                                                                         angular_{0},
                                                                         l_scale_{1.0},
                                                                         a_scale_{1.0},
                                                                         tauInicial{0.5},
                                                                         alphaValue{0.4},
                                                                         maxSimTime{maxSTime},
                                                                         ticsRate{tRate},
                                                                         actuatorValues{NUM_ACTUATORS, 0.0}
{
   /* +2 para dar la distancia a la meta como entrada. */
   sensorValues.resize(NUM_RAYS * NUM_SONARS + 2);

   prev_x = 0;
   prev_y = 0;
   deltaX = 100;
   deltaY = 100;
   deltaAngle = 3.1459;
   stuck = false;
   stuckCounter = 0;
   closestHitDistance = 100;
   nh_.param("scale_angular", a_scale_, a_scale_);
   nh_.param("scale_linear", l_scale_, l_scale_);
   is_driving = false;

   // inicilización
   stop_driving = false;

   vel_pub_ = nh_.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
   odom_sub_ = nh_.subscribe("odom", 5, &SimulationController::checkModelPosition, this);
   clock_sub_ = nh_.subscribe("clock", 1, &SimulationController::checkSimulationTime, this);
   sonar_l_sub_ = nh_.subscribe("arlo/laser/scan_left", 100, &SimulationController::checkSonarLeftValues, this);
   sonar_c_sub_ = nh_.subscribe("arlo/laser/scan_center", 100, &SimulationController::checkSonarCenterValues, this);
   sonar_r_sub_ = nh_.subscribe("arlo/laser/scan_right", 100, &SimulationController::checkSonarRightValues, this);
   service = nh_.advertiseService("evaluate_driver", &SimulationController::evaluateDriver, this);
   srvControl_robot = nh_.advertiseService("control_robot", &SimulationController::control_robot, this);
}

SimulationController::~SimulationController() {}

void SimulationController::control_sound(string name)
{
   ros::NodeHandle n;
   play_song = n.serviceClient<arlo_controller_dmaking::play_song>("play_song");

   arlo_controller_dmaking::play_song param;
   param.request.name = name;

   if (play_song.call(param))
   {
      if (name.find("motor_play") == std::string::npos)
         cout << "play_song " << name << endl;
   }
   else
   {
      cerr << "Error al reproducir el sonido " << name << endl;
   }
}

void SimulationController::simulation_color(string ctl_color)
{
   ros::NodeHandle n;
   change_color = n.serviceClient<user_interaction::change_color>("change_color");

   user_interaction::change_color param;
   param.request.color = ctl_color;

   if (change_color.call(param))
   {
      cout << "change_color " << ctl_color << endl;
   }
   else
   {
      cerr << "Error al cambiar el color de la ventana de estimulacion" << ctl_color << endl;
   }
}

void SimulationController::setDriver(ArloDriver *driver)
{
   aDriver = driver;
}

bool SimulationController::control_robot(arlo_controller_dmaking::control_robot::Request &req,
                                         arlo_controller_dmaking::control_robot::Response &res)
{
   string control = req.control;

   res.is_driving = is_driving;
   res.max_time = maxSimTime;

   if (control.find("stop") != std::string::npos)
   {
      cout << "Deteniendo robot" << endl;
      stop_driving = true;
      // arloState.hasTimeRunOut = true;
   }

   if (control.find("pause") != std::string::npos)
   {
      cout << "Pausando robot" << endl;
      // se detiene la simulación
      std_srvs::Empty gazeboParams;
      ros::service::call("/gazebo/pause_physics", gazeboParams);
      cout << "pause gazebo" << endl;
      // se pausa el sonido (incluyendo el de fondo)
      control_sound("pause_world");
      cout << "pause sound" << endl;
   }

   if (control.find("unpause") != std::string::npos)
   {
      cout << "Conduciedo el robot" << endl;
      // se continua la simulación
      std_srvs::Empty gazeboParams;
      ros::service::call("/gazebo/unpause_physics", gazeboParams);

      // se reproduce el sonido (incluyendo el de fondo)
      if (is_driving)
      {
         control_sound("unpause_world");
      }
      else
      {
         control_sound("unpause_world_min_volume");
      }
   }

   return true;
}

// void SimulationController::drive()
// {
// }

bool SimulationController::evaluateDriver(arlo_controller_dmaking::EvaluateDriver::Request &req,
                                          arlo_controller_dmaking::EvaluateDriver::Response &res)
{
   // TODO: Revisar que Gazebo este corriendo
   string pesos(req.weightsfile);
   cout << "pesos: " << req.weightsfile << endl;
   aDriver->setParameters(pesos.c_str(), req.num_ind);

   startSimulation(aDriver, req.maxtime, req.touchthreshold);

   res.time = arloState.distanceTravelled; // arloState.finishTime;
   res.dist2go = arloState.distanceToGo;
   res.damage = arloState.robotDamage;
   res.energy = arloState.distanceTravelled; // Energía usada.
   res.risk = arloState.crashingRisk;
   res.contactSpeed = fabs(arloState.finalSpeed);

   return true;
}

SimulationState SimulationController::startSimulation(ArloDriver *aDriver, int maxtime, double t_threshold)
{
   // string pesos(req.weightsfile);
   // aDriver->setParameters(pesos.c_str());

   puts("\nStarting the simulation of a new driver...");
   cerr << "---------------------------" << endl;

   std_srvs::Empty gazeboParams;
   ros::service::call("/gazebo/reset_simulation", gazeboParams);
   // inicia la musica del ambiente (incluye motor)
   control_sound("unpause_world");
   // se quita el color de la ventana de efectos
   simulation_color("hide");
   ros::Duration(0.1).sleep(); // sleep for half a second

   /* Maximum simulation time allowed */
   maxSimTime = maxtime;
   touchThreshold = t_threshold;

   ros::Rate loop_rate(40); // Frecuencia Hz con la que le robot debe tomar una decisión.

   stop_driving = false;
   linear_ = angular_ = 0;

   arloState.resetState();
   stuckCounter = 0;
   closestHitDistance = 100;
   newTau = tauInicial;

   int iter = 0;
   while (ros::ok() && showMustGoOn() && !stop_driving)
   {
      is_driving = true;
      // Send sensor values to the driver and get its answer to move the robot.
      aDriver->driveArlo(sensorValues, actuatorValues);

      if (iter++ % 10 == 0)
      {
         // sonido de los servomotores
         control_sound("motor_play");
         // string distMsg = fmt::format("D,A=[{:6.2f}, {:6.2f}] ", sensorValues[3*NUM_RAYS], sensorValues[3*NUM_RAYS+1]);
         // // printR_C(8, 50, distMsg, 32);
         // cout << distMsg << endl;
         // string outMsg = fmt::format("L,A=[{:6.2f}, {:6.2f}] ", actuatorValues[0], actuatorValues[1]);
         // // printR_C(9, 50, outMsg, 32);
         // cout << outMsg << endl;
      }
      //     << ", angular= " << actuatorValues[1] << "\n" << endl;

      linear_ = actuatorValues[0];
      angular_ = actuatorValues[1];

      // Set values in the twist object to send the actuator values to the robot in Gazebo.
      geometry_msgs::Twist twist;
      twist.angular.z = a_scale_ * angular_;
      twist.linear.x = l_scale_ * linear_;
      vel_pub_.publish(twist); // Publish the event for the twist plugin.

      ros::spinOnce();
      loop_rate.sleep();
   }

   cerr << "\033[2J";
   is_driving = false;
   stop_driving = false;

   // // se detiene el robot
   // geometry_msgs::Twist twist;
   // twist.angular.z = 0;
   // twist.linear.x = 0;
   // vel_pub_.publish(twist);
   // ros::spinOnce();

   // estados finales
   computeSimStats();

   ros::Duration(1.).sleep(); // sleep for half a second

   // se reinicia la simulación
   ros::service::call("/gazebo/reset_simulation", gazeboParams);

   // solo se deteiene el sonido del motor, los efectos de sonido se mantienen
   // control_sound("motor_stop");

   // se deteienen los efectos de sonido (incluido el motor)
   control_sound("stop_effects");
   // se ocultan los afectos
   simulation_color("hide");

   return arloState;
}

bool SimulationController::showMustGoOn()
{
   return (arloState.hasTimeRunOut == false &&
           arloState.targetReached == false);
}

void SimulationController::computeSimStats()
{

   arloState.finishTime = arloState.currentTime;
   arloState.crashingRisk = 1.0 / (closestHitDistance + 1);

   cerr << "hasTimeRunOut= " << arloState.hasTimeRunOut << " -> tiempo:" << arloState.currentTime << ", max= " << maxSimTime << "\n";
   cerr << "targetReached= " << arloState.targetReached << "\n";

   if (arloState.targetReached == true)
   {
      fmt::print(stderr, fg(fmt::color::green), "\nArlo has NAILED it :)\n");
      cerr << "finishTime = " << arloState.finishTime << "\n";
      cerr << "minDist2Go = " << arloState.distanceToGo << " <= ";
      cerr << touchThreshold << " (threshold)"
           << "\n";
      arloState.robotEnergy = 100;

      // se muestra el efecto
      simulation_color("green");
      // efecto de aplausos
      control_sound("sucess");
   }
   else if (arloState.hasTimeRunOut == true)
   {
      fmt::print(stderr, fg(fmt::color::red), "\nArlo has failed :(\n");
      cerr << "minDistance2Go = " << arloState.distanceToGo << "\n";
      cerr << "currentTime    = " << arloState.currentTime << "\n";

      // se muestra el efecto
      simulation_color("red");
      // efecto de abucheo
      control_sound("loss");
   }

   if (arloState.stuck == true)
   {
      fmt::print(stderr, fg(fmt::color::red), "\n---->>> ATASCADO <<<-----\n");
      cerr << " \t Counter = " << stuckCounter << endl;

      // se muestra el efecto
      simulation_color("red");
      // efecto de abucheo
      control_sound("loss");
   }

   cerr << "x = " << arloState.position[0] << ", y = " << arloState.position[1] << endl;
   cerr << "Contact SPEED   = " << arloState.finalSpeed << endl;
   cerr << "Crash Risk      = " << arloState.crashingRisk << endl;
   // cerr << "d2Go            = " << arloState.distanceToGo << endl;
   // err << "Dist. travelled = " << arloState.distanceTravelled << endl;

   //	res.time = arloState.finishTime;
   //   res.dist2go = arloState.distanceToGo;
   //	res.damage = arloState.robotDamage ;
   //	res.energy = arloState.distanceTravelled;
}

/*
 * Next three methods update sensor values periodically. That is, each time
 * the Platform node publishes new sensor values, these method are triggered
 * to store sensor values on the vector sensorValues.
 */

/* Method that update values from the Left Sensor. */
void SimulationController::checkSonarLeftValues(const sensor_msgs::LaserScan::ConstPtr &msg)
{
   // ROS_INFO("Tamano ranges left= %lu", msg->ranges.size());

   // cerr << "sensor= " << msg->ranges[0] << endl;
   for (int i = 0; i < msg->ranges.size(); ++i)
   {
      sensorValues[i + 0 * NUM_RAYS] = (msg->ranges[i] <= 3.0) ? (3.0 - msg->ranges[i]) : 0.0; // 0 para el sensor izq.
      if (msg->ranges[i] < closestHitDistance && arloState.distanceToGo > 1)
         closestHitDistance = msg->ranges[i];
   }
   // cerr << "sensor= " << sensorValues[0] << endl;
}

/* Method that update values from the Center Sensor. */
void SimulationController::checkSonarCenterValues(const sensor_msgs::LaserScan::ConstPtr &msg)
{
   // ROS_INFO("Tamano ranges center= %lu", msg->ranges.size());

   for (int i = 0; i < msg->ranges.size(); ++i)
   {
      sensorValues[i + 1 * NUM_RAYS] = (msg->ranges[i] <= 3.0) ? (3.0 - msg->ranges[i]) : 0.0; // 1 para el sensor central.
      if (msg->ranges[i] < closestHitDistance && arloState.distanceToGo > 1)
         closestHitDistance = msg->ranges[i];
   }
}

/* Method that update values from the Right Sensor. */
void SimulationController::checkSonarRightValues(const sensor_msgs::LaserScan::ConstPtr &msg)
{
   // ROS_INFO("Tamano ranges right= %lu", msg->ranges.size());

   for (int i = 0; i < msg->ranges.size(); ++i)
   {
      sensorValues[i + 2 * NUM_RAYS] = (msg->ranges[i] <= 3.0) ? (3.0 - msg->ranges[i]) : 0.0; // 2 para el sensor der.
      if (msg->ranges[i] < closestHitDistance && arloState.distanceToGo > 1)
         closestHitDistance = msg->ranges[i];
   }

   // sensorValues[3*NUM_RAYS+0] = deltaX;
   // sensorValues[3*NUM_RAYS+1] = deltaY;
   sensorValues[3 * NUM_RAYS + 0] = arloState.distanceToGo - 1.4;
   sensorValues[3 * NUM_RAYS + 1] = deltaAngle;
}

/* This method periodically updates the current position of the platform */
void SimulationController::checkModelPosition(const nav_msgs::Odometry::ConstPtr &msg)
{
   if (arloState.targetReached == true)
      return;

   // ROS_INFO("Seq: [%d]", msg->header.seq);
   // ROS_INFO("Position-> x: [%f]", msg->pose.pose.position.x);
   //  ROS_INFO("Vel-> Linear: [%f], Angular: [%f]",
   //  		msg->twist.twist.linear.x,
   //    	   msg->twist.twist.angular.z);
   //  arloState.currentPosition = msg->pose.pose.position.x;
   //  if (arloState.currentPosition >= goalDistance)   // Esta es la distancia del pasillo en Gazebo.
   //  	arloState.finishLineCrossed = true;

   tf2::Quaternion quat_dir;

   quat_dir.setX(msg->pose.pose.orientation.x);
   quat_dir.setY(msg->pose.pose.orientation.y);
   quat_dir.setZ(msg->pose.pose.orientation.z);
   quat_dir.setW(msg->pose.pose.orientation.w);

   double roll, pitch, yaw;
   tf2::Matrix3x3(quat_dir).getRPY(roll, pitch, yaw);

   // cout << "roboto dir.= " << yaw*180 / M_PI << endl;

   // msg->pose.pose.orientation.
   /* Calcular la velocidad promedio usando el promedio exponencial (darle más peso a valores recientes de velocidad) */
   newTau = alphaValue * msg->twist.twist.linear.x + (1 - alphaValue) * newTau;
   arloState.finalSpeed = newTau;
   // cout << "\nvelocidad = " << newTau << endl;

   /* Obtener la posición actual de robot. */
   double new_x = msg->pose.pose.position.x;
   double new_y = msg->pose.pose.position.y;

   // Compute differente between target and robot in each dimension (kinda distance)
   deltaX = 0.0 - new_x;
   deltaY = 0.0 - new_y;
   // Angle in format (0, PI), (0,-PI)
   double targetDirection = atan(fabs(deltaY / deltaX));
   if (deltaX < 0 && deltaY > 0)
   { // 2o cuadrante -,+
      targetDirection = M_PI - targetDirection;
      // cerr << "\n2o: ";
   }
   else if (deltaX < 0 && deltaY < 0)
   { // 3er cuadrante -,-
      targetDirection = -M_PI + targetDirection;
      // cerr << "\n3er: ";
   }
   else if (deltaX > 0 && deltaY < 0)
   { // 4o cuadrante +,-
      targetDirection = -targetDirection;
      // cerr << "\n4o: ";
   }

   deltaAngle = atan2(sin(targetDirection - yaw), cos(targetDirection - yaw));
   // cout << "targetAng= " << targetDirection*180 / M_PI << endl;
   // cout << "deltaAngle= " << deltaAngle*180 / M_PI << endl;

   // arloState.currentPosition = msg->pose.pose.position.x;
   arloState.position[0] = new_x;
   arloState.position[1] = new_y;
   arloState.distanceToGo = dist2Go(new_x, new_y);

   // Como los mensajes de la posición actual están en una cola, es posible que al comenzar
   // una nueva simulación, en la cola se hayan quedado valores de una posición cercana a
   // la meta. De esta manera, es posible que al inicio de una simulación, de manera incorrecta
   // se detecte que el robot llegó a la meta.
   // Cuando esto pasa, el tiempo es casi cero, así que se puede usar esto para ignorar algún
   // posible falso positivo. Si el tiempo es menor que 2 y se alcanzó el objeto, quiere decir que
   // se está leyendo un valor de la cola y no la posición actual.
   // cout << "touchThreshold " << touchThreshold << "   " << arloState.currentTime << endl;
   if (arloState.distanceToGo <= touchThreshold && arloState.currentTime > 1)
   {
      arloState.targetReached = true;
      cerr << "\n\nTARGET REACHED: " << arloState.distanceToGo << endl;
      return;
   }

   /* Determinar si el robot ya se quedó atorado. */
   /* Si la distancia que ha avanzado es menor que un umbral pequeño */
   double distanceBefore = arloState.distanceTravelled;

   // Compute the new distance travelled.
   arloState.distanceTravelled += distance(prev_x, prev_y, new_x, new_y);

   // Check if the robot has moved or not.
   if (abs(distanceBefore - arloState.distanceTravelled) < 0.001)
   {
      stuckCounter++;
      // cout << "Stuck counter: " << stuckCounter << "\n";
      if (stuckCounter > 280)
      {
         arloState.stuck = true;
         arloState.hasTimeRunOut = true;
      }
   }
   else
      stuckCounter = 0;

   // Swap values for next iteration.
   prev_x = new_x;
   prev_y = new_y;
}

/* Calcula la distancia de la plataforma (x,y) al punto meta. */
double SimulationController::dist2Go(double x, double y)
{
   /* En el escenario sencillo, suponemos que la meta está en el punto 0,0 */
   double xTarget = -2.013;
   double yTarget = 2.586;

   // cout << "x robot = " << x << ", y robot" << y << endl;
   double distToGo = distance(xTarget, yTarget, x, y);
   // cout << "distance" << distToGo << " x" << x << " y" << y << endl;
   return distToGo;
}

double SimulationController::distance(double x1, double y1, double x2, double y2)
{
   double sum = (x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2);
   return sqrt(sum);
}

void SimulationController::checkSimulationTime(const rosgraph_msgs::Clock::ConstPtr &msg)
{
   arloState.currentTime = msg->clock.toSec();
   // cerr << "\nTiempo simulacion: " <<  msg->clock.toSec() << ", max= " << maxSimTime << endl;

   if (arloState.currentTime >= maxSimTime)
   {
      arloState.hasTimeRunOut = true;
      // cerr << "\nFIN: tiempo simulacion: " <<  arloState.currentTime << ", max= " << maxSimTime << endl;
   }
}

int SimulationController::getNumSensors()
{
   return NUM_RAYS * NUM_SONARS;
}

int SimulationController::getNumActuators()
{
   return NUM_ACTUATORS;
}

int SimulationController::getNumInputs()
{
   return getNumSensors() + 2; // Delta X y Delta Y
}

void SimulationController::printR_C(int ren, int col, string text, int color)
{
   cerr << "\033 7";
   cout << "\033[38;5;" << color << "m";
   cout << "\033[" << ren << ";" << col << "H";
   cout << text << endl;
   cerr << "\033 8" << endl;
}

File: src/arlo_controller_dmaking/scripts/simulationController.py
#!/usr/bin/env python3


import rospy
import numpy as np
from arlo_controller_dmaking.srv import EvaluateDriver, EvaluateDriverResponse
from nav_msgs.msg import Odometry
from sensor_msgs.msg import LaserScan
from rosgraph_msgs.msg import Clock as Cl
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
from sensor_msgs.msg import LaserScan
from std_srvs.srv import Empty
from gazebo_msgs.srv import GetModelState
from arlo_controller_dmaking.srv import play_song as PlaySong
from arlo_controller_dmaking.srv import control_robot as ControlRobot
from arlo_controller_dmaking.srv import control_robotResponse
from user_interaction.srv import change_color
import tf
import math
from tensorflow import convert_to_tensor

def resetState():
    return {"timeRunOut": False, 
            "stuck": False,
            "finishLineCrossed": False, 
            "finishTime": 0,
            "targetReached": False, 
            "crashingRisk":0,
            "robotEnergy":100,
            "robotDamage":0,
            "position":[0,0,0],
            "distanceToGo":1000,
            "distanceTravelled":0,
            "currentTime":0}
    
simulationState = resetState()


class simulationController:
    NUM_RAYS = 16
    n_aux  = 0
    is_driving = False
    stop_driving = False

    maxSimTime = 300
    touchThreshold = 0
    stuckCounter = 0
    closestHitDistance = 100
    newTau = 0
    tauInicial = 0.5
    alphaValue = 0.4
    prev_x = 0
    prev_y = 0
    deltaAngle = .1459

    linear_ = 0
    angular_ = 0
    l_scale_ = 1
    a_scale_ = 1
    sensorValues = np.zeros(NUM_RAYS * 3 + 2) # guarda los valores de los sensores y delta x,y
    driver = None
    vel_pub_ = None

    def __init__(self) :
        # connect to topic for control robot
        self.vel_pub_ = rospy.Publisher("/cmd_vel", Twist, queue_size=1)
        odom_sub_ = rospy.Subscriber("odom", Odometry, self.checkModelPosition)
        clock_sub_ = rospy.Subscriber("clock", Cl,self.checkSimulationTime)
        sonar_l_sub_ = rospy.Subscriber("arlo/laser/scan_left", LaserScan,self.checkSonarLeftValues)
        sonar_c_sub_ = rospy.Subscriber("arlo/laser/scan_center", LaserScan,self.checkSonarCenterValues)
        sonar_r_sub_ = rospy.Subscriber("arlo/laser/scan_right", LaserScan,self.checkSonarRightValues)
        service = rospy.Service("evaluate_driver", EvaluateDriver,self.evaluateDriver)
        srvControl_robot = rospy.Service("control_robot", ControlRobot,self.control_robot)
        
    def setDriver(self, driver):
        self.driver = driver
        self.sensorValues = np.zeros(driver.getNumInputs())


    def simulation_color(self, ctl_color):
        try:
            srv = rospy.ServiceProxy("change_color", change_color)
            srv(ctl_color)
        except Exception as e:
            print(f"Error al cambiar el color de la ventana de estimulacion:\n{e}")
           
            
    def startSimulation(self, driver, maxtime, t_threshold):
        global simulationState
        print("\nStarting the simulation of a new driver...")
        
        # reset simulation
        srv_r = rospy.ServiceProxy("/gazebo/reset_simulation", Empty)
        srv_r()
        # play sound (engine sound included)
        self.control_sound("unpause_world")
        # stop stimulation of simulation result
        self.simulation_color("hide")
        # sleep before to start simulation
        t = rospy.Duration(0.1)
        rospy.sleep(t)
        # maximum simulation time allowed
        self.maxSimTime = maxtime
        self.touchThreshold = t_threshold

        # Frecuencia Hz con la que le robot debe tomar una decisión
        rate = rospy.Rate(40)
        self.stop_driving = False
        self.linear_ = self.angular_ = 0

        srv = rospy.ServiceProxy("/gazebo/reset_simulation", Empty)
        srv()

        # reset states of robot
        simulationState = resetState()

        self.stuckCounter = 0
        self.closestHitDistance = 100
        self.newTau = self.tauInicial

        iter = 0
        actuatorValues = [0,0]
        history_sensor_values = []
        
        while not rospy.is_shutdown() and self.showMustGoOn() and not self.stop_driving:
        
            self.is_driving = True
            
            # calculate the output to control the neurocontroller
            actuatorValues = driver.driveArlo([self.sensorValues])
            history_sensor_values.append(self.sensorValues) 
      
            if iter % 10 == 0:
                self.control_sound("motor_play")

            self.linear_ = actuatorValues[0]
            self.angular_ = actuatorValues[1]
    
            self.values_to_neurocontroller(self.a_scale_ * self.angular_, self.l_scale_ * self.linear_)

            rate.sleep()
            # print(self.a_scale_ * self.angular_, self.l_scale_ * self.linear_)
        
        self.is_driving = False
        self.stop_driving = False

        # dir of values
        
        # save sensor values
        obj_dir = f"{problem_dir}/obj_space_gen_{str_num}.out"
        f = open(obj_dir, "a")

        str_obj_space = " ".join(res_obj_space.astype(str))
        f.write(f"{str_obj_space}\n")
        f.close()

        # calculate the state of robot
        self.computeSimStats()

        # low velocity to stop robot
        self.values_to_neurocontroller(0, actuatorValues[0]*0.5)
        t = rospy.Duration(1)
        rospy.sleep(t)

        # stop robot
        self.values_to_neurocontroller(0,0)

        # stop stimulations
        self.control_sound("stop_effects")
        self.simulation_color("hide")

        srv = rospy.ServiceProxy("/gazebo/reset_simulation", Empty)
        srv()
        

    def computeSimStats(self):
        global simulationState

        simulationState["finishTime"] = simulationState["currentTime"]
        simulationState["crashingRisk"] = 1/(self.closestHitDistance + 1)

        if simulationState["targetReached"] == True:
            simulationState["robotEnergy"] = 100

            self.simulation_color("green");
            # success sound effect when completing the task
            self.control_sound("sucess")
        elif simulationState["timeRunOut"] == True:
            print((f"--- computeSimStats()  timeRunOut == True"))
            self.simulation_color("red")
            self.control_sound("loss")

        if simulationState["stuck"] == True:
            self.simulation_color("red")
            self.control_sound("loss")

    def values_to_neurocontroller(self,angular_z, linear_x):
        twist = Twist()
        twist.angular.z = angular_z
        twist.linear.x = linear_x
        self.vel_pub_.publish(twist)

    def showMustGoOn(self):
        global simulationState
        return simulationState["timeRunOut"] == False and simulationState["targetReached"] == False


    def evaluateDriver(self, req):
        global simulationState

        print("evaluateDriver+++++++++++++++++++++++")

        # load ANN
        self.driver.load_weights(req.weightsfile, req.num_ind)
        
        self.startSimulation(self.driver, req.maxtime, req.touchthreshold)

        return EvaluateDriverResponse(simulationState["distanceTravelled"],
                                      simulationState["distanceToGo"],
                                      simulationState["robotDamage"],
                                      simulationState["distanceTravelled"],
                                      simulationState["crashingRisk"],
                                      abs(simulationState["finalSpeed"]))

    def control_sound(self, name):
    
        try:
            play_song = rospy.ServiceProxy("play_song", PlaySong)
            play_song(name)
        except Exception as e:
            print(f"Error al reproducir el sonido:\n{e}")


    def control_robot(self, req):
        control = req.control

        if (control.find("stop") >= 0):
            print("Deteniendo robot")
            self.stop_driving = True


        if (control.find("pause") >= 0):
            # se detiene la simulación
            rospy.ServiceProxy("/gazebo/pause_physics", Empty)
            print("pause gazebo")
            # se pausa el sonido (incluyendo el de fondo)
            self.control_sound("pause_world")
            print("pause sound")


        if (control.find("unpause") >= 0):
            print("Conduciedo el robot")
            # se continua la simulación
            try:
                srvgz = rospy.ServiceProxy("/gazebo/unpause_physics", Empty)
                srvgz()
            except Exception as e:
                print(f"Error al detener la simulacion:\n{e}")

            # se reproduce el sonido (incluyendo el de fondo)
            if (self.is_driving):
                self.control_sound("unpause_world")
            else:
                self.control_sound("unpause_world_min_volume")
    

        return control_robotResponse(self.is_driving, self.maxSimTime)



    def get_model_position(self, model_name):
        rospy.wait_for_service("/gazebo/get_model_state")
        srv = rospy.ServiceProxy("/gazebo/get_model_state", GetModelState)
        res = srv(model_name, "")

        if res.status_message.find("not exist") >= 1:
            raise Exception(f"No se encuentra el modelo: {model_name}")
        
        return res.pose.position


    def checkModelPosition(self, msg):
        global simulationState

        if simulationState["targetReached"]:
            print(f"targetReached: {simulationState['targetReached']}")
            return
        
        quaternion = (
            msg.pose.pose.orientation.x,
            msg.pose.pose.orientation.y,
            msg.pose.pose.orientation.z,
            msg.pose.pose.orientation.w)
        
        roll, pitch, yaw = tf.transformations.euler_from_quaternion(quaternion)

        # Calcular la velocidad promedio usando el promedio exponencial (darle más peso a valores recientes de velocidad) 
        self.newTau = self.alphaValue * msg.twist.twist.linear.x + (1 - self.alphaValue) * self.newTau
        simulationState["finalSpeed"] = self.newTau

        #  Obtener la posición actual de robot
        new_x = msg.pose.pose.position.x
        new_y = msg.pose.pose.position.y

        # get position of the target
        pos = self.get_model_position("mesita_meta")

        # Compute differente between target and robot in each dimension (kinda distance)
        deltaX = pos.x - new_x
        deltaY = pos.y - new_y

        #  Angle in format (0, PI), (0,-PI)
        targetDirection = math.atan(abs(deltaY / deltaX))
        if (deltaX < 0 and deltaY > 0):
        # 2o cuadrante -,+z
            targetDirection = math.pi - targetDirection
        
        elif (deltaX < 0 and deltaY < 0):
        #  3er cuadrante -,-
            targetDirection = -math.pi + targetDirection
            
        elif (deltaX > 0 and deltaY < 0):
        #  4o cuadrante +,-
            targetDirection = -targetDirection
            
        self.deltaAngle = math.atan2(math.sin(targetDirection - yaw), math.cos(targetDirection - yaw))
        
        simulationState["position"][0] = new_x
        simulationState["position"][1] = new_y
        simulationState["distanceToGo"] = math.dist([pos.x, pos.y],[new_x, new_y])

        # # Como los mensajes de la posición actual están en una cola, es posible que al comenzar
        # # una nueva simulación, en la cola se hayan quedado valores de una posición cercana a
        # # la meta. De esta manera, es posible que al inicio de una simulación, de manera incorrecta
        # # se detecte que el robot llegó a la meta.
        # # Cuando esto pasa, el tiempo es casi cero, así que se puede usar esto para ignorar algún
        # # posible falso positivo. Si el tiempo es menor que 2 y se alcanzó el objeto, quiere decir que
        # # se está leyendo un valor de la cola y no la posición actual.
        if (simulationState["distanceToGo"] <= self.touchThreshold and simulationState["currentTime"] > 1):
            print(f"simulationState['currentTime']: {simulationState['currentTime']} > 1")
            simulationState["targetReached"] = True
            return;
        

        # /* Determinar si el robot ya se quedó atorado. */
        # /* Si la distancia que ha avanzado es menor que un umbral pequeño */
        distanceBefore = simulationState["distanceTravelled"]

        # # Compute the new distance travelled.
        simulationState["distanceTravelled"] += math.dist([self.prev_x, self.prev_y], [new_x, new_y])

        # # Check if the robot has moved or not.
        if (abs(distanceBefore - simulationState["distanceTravelled"]) < 0.001):
            self.stuckCounter+=1
            
            if (self.stuckCounter > 180):
                print((f"--- stuckCounter : {self.stuckCounter}"))
                simulationState["stuck"] = True
                simulationState["timeRunOut"] = True
            
        else:
            self.stuckCounter = 0

        # # Swap values for next iteration.
        self.prev_x = new_x
        self.prev_y = new_y

    def checkSimulationTime(self,msg):
        global simulationState

        simulationState["currentTime"] = msg.clock.to_sec();
        if (simulationState["currentTime"] >= self.maxSimTime):
                
            if(self.n_aux % 1600 == 0):
                print(f"++++ checkSimulationTime() , timeRunOut:{ simulationState['timeRunOut']} --- {simulationState['currentTime']} >= {self.maxSimTime}, self.n_aux:{self.n_aux}", )
            self.n_aux += 1
            simulationState["timeRunOut"] = True


    def checkSonarLeftValues(self,msg):
        global simulationState

        for i in range(len(msg.ranges)):
            if msg.ranges[i] <= 3.0:
                self.sensorValues[i + 0 * self.NUM_RAYS] = 3.0 - msg.ranges[i]
            else:
                self.sensorValues[i + 0 * self.NUM_RAYS] = 0.0

            if (msg.ranges[i] < self.closestHitDistance and simulationState["distanceToGo"] > 1):
                self.closestHitDistance = msg.ranges[i]

    def checkSonarCenterValues(self, msg):
        global simulationState

        for i in range(len(msg.ranges)):
            if msg.ranges[i] <= 3.0:
                self.sensorValues[i + 1 * self.NUM_RAYS] = 3.0 - msg.ranges[i]
            else:
                self.sensorValues[i + 1 * self.NUM_RAYS] = 0.0
             
            if (msg.ranges[i] < self.closestHitDistance and simulationState["distanceToGo"] > 1):
                self.closestHitDistance = msg.ranges[i]

    def checkSonarRightValues(self, msg):
        global simulationState

        for i in range(len(msg.ranges)):
            if msg.ranges[i] <= 3.0:
                self.sensorValues[i + 2 * self.NUM_RAYS] = 3.0 - msg.ranges[i]
            else:
                self.sensorValues[i + 2 * self.NUM_RAYS] = 0.0
             
            if (msg.ranges[i] < self.closestHitDistance and simulationState["distanceToGo"] > 1):
                self.closestHitDistance = msg.ranges[i]

        self.sensorValues[3 * self.NUM_RAYS + 0] = simulationState["distanceToGo"] - 1.4
        self.sensorValues[3 * self.NUM_RAYS + 1] = self.deltaAngle



